File: Medium/279-PerfectSquares.cpp
/*
   @author
   Akshat Atray
   Leetcode- https://leetcode.com/akshatatray/
   LeetCode Problem 279. Perfect Squares
   Language C++
   Level Easy
*/

class Solution {
public:
    int numSquares(int n) {
        static vector<int> dp {0};
        while (dp.size() <= n) {
            int m = dp.size(), sq = INT_MAX;
            for (int i=1; i*i<=m; ++i)
                sq = min(sq, dp[m-i*i]+1);
            dp.push_back(sq);
        }
        return dp[n];
    }
};

File: Medium/279-PerfectSquares.go
/*
   @author
   Akshat Atray
   Leetcode- https://leetcode.com/akshatatray/
   LeetCode Problem 279. Perfect Squares
   Language Go
   Level Easy
*/

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func numSquares(n int) int {
	var dp []int
	dp = append(dp, 0)
	for len(dp) <= n {
		m := len(dp)
		sq := 1<<7 - 1
		for i := 1; i*i <= m; i++ {
			sq = min(sq, dp[m-i*i]+1)
		}
		dp = append(dp, sq)
	}
	return dp[n]
}